### ISSUE TITLE
Samtools stats / samtools idxstats

### ISSUE THREAD

**Issue by ewels** [ewels/MultiQC#251]
Most people seem to run `samtools stats` for a consistent view of the alignment scores for a dataset. It would be nice to have a simple bar plot showing this information; the current beeswarm plot shows a lot of information and it's not super quick to understand what the alignment rate is.

Need to cherry pick categories and remove them from beeswarm. Would be good to write in a check to make sure that we get the correct percentages too.

See #250 for some discussion.

**Comment by ewels** [comment]
Done in #257

Only mapped and unmapped at the moment, but that's probably fine.

**Comment by tantrev** [comment]
Just a request: it would be quite nice if MultiQC had support for samtools' "idxstats" and "flagstat" commands. :P

**Comment by ewels** [comment]
Hi @tantrev,

There is already support for `stats` and `flagstat` in v0.8dev. Just grab a development copy of the code, though I'm hoping to release v0.8 relatively soon too.

I have previously thought about adding support for `idxstats` but I'm not sure how to visualise it properly. It's the number of aligned reads per chromosome, right?

PHil

**Comment by tantrev** [comment]
Awesome, thank you!  I'll grab a copy right now.

And yep, you got it.  I was more just being lazy and wanted to use `idxstats `for parsing but I'd imagine a parallel coordinates plot could do the job?

**Comment by ewels** [comment]
Hmm, interesting idea. I could probably trick a regular MultiQC line graph into looking like a parallel coordinates plot - could look good!

**Comment by ewels** [comment]
Damn, next problem is finding output from `idxstats`. Goes to standard out, so no common file name. Doesn't really have _any_ identifying information within the file:
```
1       249250621       860208  1
2       243199373       1106007 2
3       198022430       545145  1
4       191154276       359862  0

[...]

GL000225.1      211173  0       0
GL000192.1      547496  0       0
NC_007605       171823  9       0
hs37d5  35477943        817     0
*       0       0       4
```

Can't really just parse any files that contain four columns. Might have to just rely on users putting the phrase `idxstats` somewhere in the file path in this case 😞

**Comment by tantrev** [comment]
A small price to pay for standardized convenience. 👍   And absolutely, a hacked line graph could definitely be a winner.

**Comment by ewels** [comment]
So what's the number we want here - number of mapped reads divided by chromosome length? Or percentage aligned? I'm guessing that reporting just the number of mapped reads will just plot the chromosome lengths...

Also, how can unaligned reads be assigned to a chromosome?

**Comment by tantrev** [comment]
For myself, I'm just interested in the raw mapped reads per chromosome.  I've been working with RNA-sequencing data where it can be interesting (more macro data than samtools more-granular `stats`), and the raw reads can also help with discerning the gender of unmarked samples.

**Comment by tantrev** [comment]
Also, I think you may be referring to the last * line of the output?  [Biostars](https://www.biostars.org/p/105863/) seems to think it's just the total number of unaligned reads.

**Comment by ewels** [comment]
Nope, every chromosome has a column for unmapped reads. From the docs:

>  Each line consists of reference sequence name, sequence length, # mapped reads and # unmapped reads.

I figured that the final `*` line was unmapped / no chromosome assigned. I guess the others must be aligned reads that failed criteria somehow or something?

Anyway, the new submodule came together pretty quickly in the end. I've just pushed the code to the v0.8dev so it should be available now. I was originally planning on spacing the points by on the x axis based on chromosome length and so on, mainly because I didn't want the plot to be swamped by tiny alt chromosomes. However, I filtered so anything with less than 0.1% mapped reads is ignored and it actually looked ok. So I've left it fairly basic for now.

I also played around with plotting %aligned and mapped/chr length but they both looked rubbish, at least with my example log file. Number of mapped reads looks to be more informative than I was expecting, so I think I'll leave it as just that for now.

Let me know what you think!

**Comment by ewels** [comment]
Hi @tantrev,

I was thinking about your use case of wanting to see counts for chromosomes X and Y to confirm gender. This is something I've thought about before too. So I just added some code that tries to recognise chromosomes named `X` and `Y`. If it finds both for a sample, it plots an additional bar graph with counts / percentages for just these chromosomes. Should make it super easy to see at a glance how the genders line up.

I only have one test file which is apparently female, but it seems to work ok for me. Would be great if you see if it works for your data.. I get this:

![image](https://cloud.githubusercontent.com/assets/465550/17950774/fc6e9012-6a22-11e6-9fa0-09692d06900c.png)

Phil

**Comment by tantrev** [comment]
Wow, that's beautiful.  Well done!  Thank you.  I'm just waiting on some calculations that should be done in the next hour or so but will throw their charts up when they're done.  Your XY chart will be especially handy because this dataset just so happens to be unlabeled. :P

**Comment by tantrev** [comment]
Looks good!  Did you put a cap on the number of parallel coordinate plots?  Mine seems to be a little fragmented - then again, I'm kind of drowning in samples. I love the X/Y visualization - I have a sneaking suspicion my dataset is unfortunately mixed after seeing it.  Also, I think the text parsing that would normally be in "multiqc_samtools_idxstats.txt" may not be working yet?

If it helps, here's some sample data:
[data.tar.gz](https://github.com/ewels/MultiQC/files/436556/data.tar.gz)

Lastly, I'm not still entirely sure what that last column is, but it seems to be affected by whether data has paired-ends or not.  Thanks again!

**Comment by ewels** [comment]
Brilliant, thanks for this example data @tantrev. Several things for me to look into here.

* No cap on the plots, so not sure what's going on here. I will investigate..
* Flat data output doesn't work in the default `tsv` format as it's a nested dict currently. Works fine with `-k yaml` or `-k json`, but I guess I should probably flatten it.
* You had enough samples to trigger the 'flat' MatPlotLib plotting mode, and it looks like that doesn't yet support categories on the X axis.

Makes sense about the paired end thing. I guess it must be unmapped mates.

**Comment by ewels** [comment]
Hi @tantrev,

I've just pushed a fix in 9dc5bb531de09f7033d41c58e8d2f5e9464791a8. After a bit of investigation I figured out that there were some quite bad problems in the logic I'd used, so it was a bit of a rewrite :) Should all work now anyway!

In this commit I also added some optional config options that you can set in your `~/.multiqc_config.yaml` file. These are documented here: http://multiqc.info/docs/#samtools

I'm feeling confident, so will close this issue now, but please let me know if it works as expected for you!

Thanks again, especially for the test data,

Phil

**Comment by ojziff** [comment]
Hi @ewels 

I am trying to use the `XY counts` in the `multiqc_samtools_idxstats.txt` to determine the gender of publically available RNAseq samples from GEO where the sample gender is not documented in the accession metadata.

Do you have a suggested threshold based on the % of Y chromosome reads to predict that the sample is from a male? It seems roughly around > 0.5% Y Chr reads is male but I wondered if you had a more accurate suggested threshold. I have some samples in the 0.5 - 1.0% region, which I'm not sure which gender they are but I'm inclined to say male? For example in this:

![Screenshot 2022-05-19 at 14 49 02](https://user-images.githubusercontent.com/41295194/169309294-79d5e72e-4d48-4374-ba60-0b2b39d92a67.png)

Many thanks!

**Comment by olechnwin** [comment]
> Hi @ewels
> 
> I am trying to use the `XY counts` in the `multiqc_samtools_idxstats.txt` to determine the gender of publically available RNAseq samples from GEO where the sample gender is not documented in the accession metadata.
> 
> Do you have a suggested threshold based on the % of Y chromosome reads to predict that the sample is from a male? It seems roughly around > 0.5% Y Chr reads is male but I wondered if you had a more accurate suggested threshold. I have some samples in the 0.5 - 1.0% region, which I'm not sure which gender they are but I'm inclined to say male? For example in this:
> 
> ![Screenshot 2022-05-19 at 14 49 02](https://user-images.githubusercontent.com/41295194/169309294-79d5e72e-4d48-4374-ba60-0b2b39d92a67.png)
> 
> Many thanks!

I'm wondering about the same. In my case, alignment to chr Y ranges from 15% - 1.3%. 
![image](https://user-images.githubusercontent.com/20069067/219677306-8ab52d43-cf60-46df-a367-16e92deb6718.png)

**Comment by ewels** [comment]
Hi @ojziff, @olechnwin,

Posting on merged pull-requests or closed issues is always a bit risky, usually better to open new for stuff like this as there's a high chance I will miss things.

But having said that, the question is a biological one rather than technical. The answer likely begins with "it depends" and will be impacted by a lot of upstream steps. I think you're better off asking somewhere like [Biostars](https://www.biostars.org/) or [Bioinformatics Stack Exchange](https://bioinformatics.stackexchange.com/) for stuff like this.

Cheers,

Phil

**Comment by olechnwin** [comment]
Hi @ewels ,

Thanks for your response. I am always hesitant to open new issue for stuff like this. I don't want to clutter the space. Thank you for your suggestion.

**Comment by ewels** [comment]
No worries! Closing issues is easy, but if I miss a comment on a closed issue you'll just never get a reply 🙂 (as happened above).

There's always the Nextflow Gitter chat if you're not sure if something is GitHub issue worthy!

**Comment by ewels** [comment]
🙈 Probably going to move to an alternative soon. Is Slack blocked?? 😅
